### PR TITLE
add a graph why command to explain why certain package appears in the graph

### LINKED ISSUE #462: add command to show the reasons a package is included in the build graph
For debugging, especially when there are conflicts, it would be useful to have a command to summarize the reasons a package appears in the build graph. It should show all paths from the top-level to all versions of the package, including all edges and the edge type.

Optional enhancements could be to:

* Filter by edge type
* Filter by package version
* Produce output in dot format suitable for rendering into a picture with graphviz

### PR DESCRIPTION
**Pull request by shubhbapna**
fixes #462 

Didn't get the bonus marks for converting to dot file : (

### CODE REVIEW

**Comment by shubhbapna** [comment]
Here are some examples:

```bash
$ fromager graph why ./graph.json dill

dill==0.3.8
   * is an install dependency of datasets==3.0.1 with req dill<0.3.9,>=0.3.0
   * is an install dependency of datasets==2.21.0 with req dill<0.3.9,>=0.3.0

dill==0.3.9
   * is an install dependency of multiprocess==0.70.16 with req dill>=0.3.8
   * is an install dependency of lm-eval==0.4.4 with req dill
   * is an install dependency of evaluate==0.4.3 with req dill
```

```bash
$ fromager graph why ./graph.json dill --recursive-depth 1

dill==0.3.8
   * is an install dependency of datasets==3.0.1 with req dill<0.3.9,>=0.3.0
     * is an install dependency of instructlab==0.19.0 with req datasets>=2.18.0
     * is an install dependency of trl==0.11.1 with req datasets
     * is an install dependency of instructlab-training==0.5.0 with req datasets>=2.15.0
     * is an install dependency of instructlab-training==0.5.0 with req datasets>=2.15.0
     * is an install dependency of lm-eval==0.4.4 with req datasets>=2.16.0
     * is an install dependency of evaluate==0.4.3 with req datasets>=2.0.0
     * is an install dependency of outlines==0.0.46 with req datasets
   * is an install dependency of datasets==2.21.0 with req dill<0.3.9,>=0.3.0
     * is an install dependency of instructlab-sdg==0.3.0 with req datasets<3.0.0,>=2.18.0

dill==0.3.9
   * is an install dependency of multiprocess==0.70.16 with req dill>=0.3.8
     * is an install dependency of datasets==3.0.1 with req multiprocess
     * is an install dependency of datasets==2.21.0 with req multiprocess
     * is an install dependency of evaluate==0.4.3 with req multiprocess
   * is an install dependency of lm-eval==0.4.4 with req dill
     * is an install dependency of instructlab-eval==0.3.1 with req lm-eval>=0.4.4
   * is an install dependency of evaluate==0.4.3 with req dill
     * is an install dependency of lm-eval==0.4.4 with req evaluate
     * is an install dependency of lm-eval==0.4.4 with req evaluate>=0.4.0
```

File: src/fromager/commands/graph.py [hunk 1]
```diff
@@ -153,6 +162,69 @@ def explain_duplicates(wkctx, graph_file):
...
 
+@graph.command()
+@click.option(
+    "--version",
+    type=clickext.PackageVersion(),
+    multiple=True,
+    help="filter by version for the given package",
+)
+@click.option(
+    "--recursive-depth",
```
Comment (dhellmann) [inline src/fromager/commands/graph.py:173]: Maybe just "depth"?

File: src/fromager/commands/graph.py [hunk 2]
```diff
@@ -153,6 +162,69 @@ def explain_duplicates(wkctx, graph_file):
...
+    type=int,
+    default=0,
+    help="recursively get why each package depends on each other. Set depth to -1 for full recursion till root",
+)
+@click.option(
+    "--req-type",
+    type=clickext.RequirementType(),
+    multiple=True,
+    help="filter by req type",
```
Comment (dhellmann) [inline src/fromager/commands/graph.py:182]: We use "req" as an abbreviation a lot internally, but in the user interface maybe we should spell out "requirement".

File: src/fromager/commands/graph.py [hunk 3]
```diff
@@ -153,6 +162,69 @@ def explain_duplicates(wkctx, graph_file):
...
+        if parent.destination_node.key == ROOT:
+            continue
+        if req_type and parent.req_type not in req_type:
+            continue
+
+        print(
+            f"{'  ' * depth} * is an {parent.req_type} dependency of {parent.destination_node.key} with req {parent.req}"
+        )
+        if recursive_depth and (recursive_depth == -1 or depth <= recursive_depth):
+            find_why(graph, parent.destination_node, recursive_depth, depth + 1, [])
```
Comment (dhellmann) [inline src/fromager/commands/graph.py:225]: If I filter based on requirement types I might not see any output of this loop. A message saying that all dependencies were filtered would help understand that case.

File: src/fromager/requirements_file.py [hunk 4]
```diff
@@ -17,6 +17,19 @@ class RequirementType(StrEnum):
     BUILD_BACKEND = "build-backend"
     BUILD_SDIST = "build-sdist"
 
+    def __eq__(self, other: object) -> bool:
```
Comment (dhellmann) [inline src/fromager/requirements_file.py:20]: This looks worth having a unit test for.

File: src/fromager/commands/graph.py [hunk 5]
```diff
@@ -153,6 +162,69 @@ def explain_duplicates(wkctx, graph_file):
...
+)
+@click.option(
+    "--recursive-depth",
+    type=int,
+    default=0,
+    help="recursively get why each package depends on each other. Set depth to -1 for full recursion till root",
+)
+@click.option(
+    "--req-type",
+    type=clickext.RequirementType(),
```
Comment (dhellmann) [inline src/fromager/commands/graph.py:180]: Does setting the type like this show the user what the valid values are? If not, can we pass the set of choices so click will report that?

**Review by dhellmann — APPROVED** [review]

**Pull request merged**